src/main/java/hu/blackbelt/structured/map/proxy/MapProxy.java

```diff
@@ -706,6 +708,80 @@ private boolean invokeEquals(Object proxy, Object[] args) throws InvocationTarge
         return false;
     }
 
+    private void invokeAdd(Method m, Object[] args) throws ExecutionException {
+        if (params.isImmutable()) {
+            throw new IllegalStateException("Could not call set on immutable object");
+        }
+        String attrName = Character.toLowerCase(m.getName().charAt(5)) + m.getName().substring(6);
+        AttributeInfo attributeInfo = typeInfoCache.get(clazz).get(attrName);
+        Object values = internal.get(getKeyName(clazz, attrName));
+
+        List<Object> valuesArray = (values instanceof Collection<?>)
+                ? valuesArray = new ArrayList<>((Collection) values)
+                : new ArrayList<>();
+
+        if (args[0] instanceof Object[]) {
+            Object[] valuesForAdd = (Object[]) args[0];
+            for (Object value : valuesForAdd) {
+                if (attributeInfo != null && attributeInfo.isComposite() && value instanceof MapHolder) {
+                   if (attributeInfo.getPropertyType().isInterface()) {
+                       MapHolder proxy = (MapHolder) MapProxy.builder(attributeInfo.getPropertyType())
+                               .withParams(params)
+                               .withMap(((MapHolder) value).toMap())
+                               .newInstance();
+                       proxy.toMap().entrySet().forEach(e -> {
+                           internal.put(e.getKey(), e.getValue());
+                       });
+                   } else {
+                       throw new IllegalArgumentException(String.format("The attribute %s in %s is not an interface. The @Embedded attributes type has to be interface.", attrName, clazz.getName()));
+                   }
+                } else {
+                    valuesArray.add(value);
+                }
+            }
+        } else {
+            valuesArray.add(args[0]);
+        }
+        internal.put(getKeyName(clazz, attrName), Collections.unmodifiableList(valuesArray));
+    }
+
+    private void invokeRemove(Method m, Object[] args) throws ExecutionException {
+        if (params.isImmutable()) {
+            throw new IllegalStateException("Could not call set on immutable object");
+        }
+        String attrName = Character.toLowerCase(m.getName().charAt(10)) + m.getName().substring(11);
+        AttributeInfo attributeInfo = typeInfoCache.get(clazz).get(attrName);
+        Object values = internal.get(getKeyName(clazz, attrName));
+
+        List<Object> valuesArray = (values instanceof Collection<?>)
+                ? valuesArray = new ArrayList<>((Collection) values)
+                : new ArrayList<>();
+
+        if (args[0] instanceof Object[]) {
+            Object[] valuesForRemove = (Object[]) args[0];
+            for (Object value : valuesForRemove) {
+                if (attributeInfo != null && attributeInfo.isComposite() && value instanceof MapHolder) {
```

Comment:
I think `invokeAdd` and `invokeRemove` can be merged, because minimal differences are presented, and can be a boolean which discriminate add or remove operation on collection.

Reply:
[b7d7975](https://github.com/BlackBeltTechnology/structured-map-proxy/pull/82/commits/b7d79756f1b886512710a06d2298613ff0d17eae)